src/minio/managers/policy_manager.py

```diff
@@ -295,6 +296,114 @@ async def _attach_detach_policy(
+                f"Invalid S3 path format: {path}. Must start with s3:// or s3a://"
+            )
+
+        # Extract bucket and path from S3 URL
+        if path.startswith("s3a://"):
+            path_without_scheme = path.replace("s3a://", "")
+        else:  # s3://
+            path_without_scheme = path.replace("s3://", "")
+        path_parts = path_without_scheme.split("/", 1)
+
+        if not path_parts:
+            raise PolicyOperationError(f"Invalid S3 path format: {path}")
+
+        bucket_in_path = path_parts[0]
+
+        # Validate bucket matches our configuration
+        # TODO: support multiple buckets in the future
+        if bucket_in_path != self.config.default_bucket:
+            raise PolicyOperationError(
+                f"Path bucket '{bucket_in_path}' does not match configured bucket '{self.config.default_bucket}'"
+            )
+
+        # Return the path part (everything after bucket)
+        if len(path_parts) <= 1:
+            raise PolicyOperationError(
+                f"S3 path must include a path component after bucket: {path}"
+            )
+
+        return path_parts[1]
+
+    def _add_path_to_list_bucket_statement(
+        self, policy_model: PolicyModel, clean_path: str
+    ) -> None:
+        """Add path prefixes to existing ListBucket statement."""
+        list_bucket_stmt = self._find_list_bucket_statement(policy_model)
+
+        if not list_bucket_stmt or not list_bucket_stmt.condition:
+            return
+
+        existing_prefixes = list_bucket_stmt.condition["StringLike"]["s3:prefix"]
+        new_prefixes = [f"{clean_path}/*", f"{clean_path}"]
+
+        for prefix in new_prefixes:
+            if prefix not in existing_prefixes:
+                existing_prefixes.append(prefix)
+
+    def _find_list_bucket_statement(
+        self, policy_model: PolicyModel
+    ) -> PolicyStatement | None:
+        """Find the ListBucket statement with prefix conditions."""
+        for stmt in policy_model.policy_document.statement:
+            if (
+                PolicyAction.LIST_BUCKET in stmt.action
+                and stmt.condition
+                and "StringLike" in stmt.condition
+                and "s3:prefix" in stmt.condition["StringLike"]
+            ):
+                return stmt
+        return None
+
+    def _add_object_level_statement(
+        self,
+        policy_model: PolicyModel,
+        clean_path: str,
+        permission_level: PolicyPermissionLevel,
+    ) -> None:
+        """Add object-level permissions statement for the path."""
+        actions = self._get_actions_for_permission_level(permission_level)
+
+        new_statement = PolicyStatement(
+            effect=PolicyEffect.ALLOW,
+            action=actions,
+            resource=[f"arn:aws:s3:::{self.config.default_bucket}/{clean_path}/*"],
+            condition=None,
+            principal=None,
+        )
+        policy_model.policy_document.statement.append(new_statement)
```

Comment:
This could add duplicate paths to the policy statement list, is that what you want?

It also could add multiple statements with different perms for the same path

Reply:
I added a todo item here - I was thinking to add a more general policy de-dup method before update action happens.

Reply:
👍

Reply:
👍

Reply:
Do you still need eq and hashcode?

Reply:
The conflict detection code seems like it could be simpler, unless I'm misunderstanding something. Is it true that each statement should deal with a unique path, and there should only be one statement per path? If so, I think you can do something like

```
for stat in statements:
    if get_path(stat) == path:
        statements.remove(stat)
statements.add(new_statement)
```

All the effect and actions checking doesn't seem necessary

Reply:
no - the resource of state can have multiple paths.

Reply:
Ok, I'm not quite understanding the resolution code in that case - IIUC if a statement has a path matching the path in the new statement, it's removed from the list and new_statement replaces it. But again, IIUC, if the old statement has multiple paths, some of those paths may not match the new path and be removed, removing permissions that the user should keep

Reply:
yea - that sounds like a bug - instead I will just raise when there were conflicts.

Reply:
yea - that sounds like a bug - instead I will just raise when there were conflicts.

Reply:
Under what circumstances could the error be thrown?

Reply:
when there are conflicts of path detected... I might not following what you are trying to ask. You asked to detect this -  I personally think `if new_statement not in policy_model.policy_document.statement:` is good.  If you don't like this `_detect_policy_conflicts` - I wanted to remove that actually.

Reply:
I do think the conflict logic could be simplified. Basically all you care about is whether there's a statement in the policy that has the same path in it.

* If it's identical to the current statement do nothing and don't add the current statement
* Based on the current logic, if it's not identical throw an error.

It seems like that could be implemented a lot more simply than the current code

Reply:
Oh, you need to set the model to frozen to get `__hash__`

Reply:
👍